File: back/model/scenario.go
package model

// Scenario holds all the information of a scenario
type Scenario struct {
	Language  string `json:"language"`
	MapImg    string `json:"mapImg"`
	AvatarImg string `json:"avatarImg"`
	Steps     []Step `json:"steps"`
}

// Step holds the information of one step of a scenario
type Step struct {
	Title       string `json:"title"`
	Text        string `json:"text"`
	MapPosition struct {
		X int `json:"x"`
		Y int `json:"y"`
	} `json:"mapPosition"`
	Template    string `json:"template"`
	Tests       []Test `json:"tests"`
	Validations []Test `json:"validations"`
	Results     struct {
		Success        Result  `json:"success"`
		PartialSuccess *Result `json:"partialSuccess,omitempty"`
		Failure        Result  `json:"failure"`
	} `json:"results"`
}

// Result defines an expected result for a step of a scenario
type Result struct {
	Finish    bool   `json:"finish"`
	Step      *int   `json:"step"`
	Score     int    `json:"score"`
	Threshold *int   `json:"threshold,omitempty"`
	Text      string `json:"text,omitempty"`
}

File: server/src/main/kotlin/com/zenika/nckotlinserver/model/Scenario.kt
package com.zenika.nckotlinserver.model

import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonInclude.Include.NON_NULL

/**
 * Holds all the information of a scenario
 */
@JsonInclude(NON_NULL)
data class Scenario(
        val language: String,
        val mapImg: String,
        val avatarImg: String,
        val steps: Array<Step>
) : Entity {
    override fun id() = language
}

/**
 * Holds the information of one step of a scenario
 */
data class Step(
        val title: String,
        val text: String,
        val mapPosition: Coord,
        val template: String,
        val tests: Array<Test>,
        val validations: Array<Test>,
        val results: Results
)

@JsonInclude(NON_NULL)
data class Results(
        val success: Result,
        val partialSuccess: Result?,
        val failure: Result
)

/**
 * Defines an expected result for a step of a scenario
 */
@JsonInclude(NON_NULL)
data class Result(
        val finish: Boolean,
        val step: Int?,
        val score: Int,
        val threshold: Int?,
        val text: String?
)